pynicotine/transfers.py

```diff
@@ -449,28 +449,32 @@ def file_limit_reached(self, user):
 
     def slot_limit_reached(self):
 
-        maxupslots = self.config.sections["transfers"]["uploadslots"]
+        uploadslots = self.config.sections["transfers"]["uploadslots"]
 
-        if maxupslots <= 0:
-            maxupslots = 1
+        if uploadslots <= 0:
+            uploadslots = 1
 
         in_progress_count = 0
-        now = time.time()
+        current_time = time.time()
 
-        for i in self.uploads:
+        for i in reversed(self.uploads):
             if i.conn is not None and i.speed is not None:
                 # Currently transferring
                 in_progress_count += 1
 
-            elif (now - i.last_status_change) < 30:
+            elif (current_time - i.last_status_change) < 30:  # ToDo: performance bottleneck
                 # Transfer initiating, changed within last 30 seconds
 
                 if (i.req is not None
                         or i.conn is not None and i.speed is None
                         or i.status == "Getting status"):
```

Comment:
When there are a large number of historic uploads to iterate, testing this condition can become very resource intensive when there are many queued transfers and a low number of slots is set. Breaking out of the loop seems to work around this, but maybe it is redundant to check this arbitrary state like this, maybe the second set of conditions are faster so they can be processed first?

Reply:
Most of this seems redundant now, I think we can simply check the status messages.
`if i.status in ("Getting status", "Establishing connection"):`

And the top one:
`if i.status == "Transferring":`